### PR TITLE
persist: implement Consensus on top of sqlite.

### PR DESCRIPTION
**Pull request by ruchirK**
This commit adds a new Consensus built on top of sqlite, which will be the
Consensus implementation while we are prototyping, and most likely also the
Consensus implementation used for on-prem deployments.

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/persist/src/sqlite.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        Ok(SqliteConsensus {
+            conn: Arc::new(Mutex::new(conn)),
+            shard,
+        })
+    }
+}
+
+#[async_trait]
+impl Consensus for SqliteConsensus {
+    async fn head(&self, _deadline: Instant) -> Result<Option<VersionedData>, Error> {
```
Comment (danhhz) [inline src/persist/src/sqlite.rs:66]: is there any way for us to use the timeout? also possibly relevant for `open`

File: src/persist/src/sqlite.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            })
+            .optional()?;
+
+        let seqno = current.as_ref().map(|c| c.seqno);
+        if seqno != expected {
+            return Ok(Err(current));
+        }
+
+        if let Some(seqno) = seqno {
+            if new.seqno <= seqno {
```
Comment (danhhz) [inline src/persist/src/sqlite.rs:117]: we could move the check (presumably reframed as something like `new.seqno <= expected`) before even talking to sqlite right? if there's some way to make that work, let's do it

File: src/persist/src/sqlite.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+             VALUES ($shard, $sequence_number, $data)",
+        )?;
+
+        insert_stmt.execute(named_params! {
+            "$shard": self.shard,
+            "$sequence_number": new.seqno.0,
+            "$data": new.data,
+        })?;
+
+        drop(insert_stmt);
```
Comment (danhhz) [inline src/persist/src/sqlite.rs:133]: hmm? what's going on here? if it's not obvious, let's leave a comment

File: src/persist/src/sqlite.rs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+#[derive(Debug)]
+pub struct SqliteConsensus {
+    conn: Arc<Mutex<Connection>>,
+    shard: String,
+}
+
+impl SqliteConsensus {
+    /// Open a sqlite-backed [Consensus] instance at `path`, for the collection
+    /// named `shard`.
+    pub fn open(path: &Path, shard: String) -> Result<Self, Error> {
```
Comment (danhhz) [inline src/persist/src/sqlite.rs:43]: nit: do the `AsRef<Path>` trick here, it seems to be idiomatic for this sort of thing

File: src/persist/src/sqlite.rs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+use rusqlite::{named_params, params, Connection, OptionalExtension};
+use tokio::sync::Mutex;
+
+use crate::error::Error;
+use crate::storage::{Consensus, SeqNo, VersionedData};
+
+const APPLICATION_ID: i32 = 0x0678_ef32; // chosen randomly
+
+const SCHEMA: &str = "
+CREATE TABLE consensus (
```
Comment (danhhz) [inline src/persist/src/sqlite.rs:26]: feels like this should have a primary key. perhaps `(shard, sequence_number)`?

**Review by danhhz — COMMENTED** [review]
Approach seems fine to me, but I don't know anything about sqlite or rusqlite. Maybe @benesch could take a look (or at least point us to someone who could)?

> I stole the APPLICATION_ID pattern from both the catalog and stash. Not sure if we want to continue doing this / if there's another, better value for the APPLICATION_ID?

Dunno

> My working assumption with this pr was "do the minimum needed to get something out the door that can work with the current Consensus API with the understanding that we will wipe out any persisted state as we iterate to the final version of the API". Is that a fair assumption?

In the short term, yes. In the new platform world, we're committing to backward compatibility much earlier than we were in the "fast restarts" world, so we're going to have to commit to something in the 2-3 month timeframe. We'd also have to do something like the old "wipe" mechanism if we e.g. want to turn the background system tables test back on in a shorter timeframe than that.

> I'm pretty sure my conditional put logic only works if sqlite transactions are serializable. Not totally sure what the defaults are there. will look into it now. Also, i suppose im conflicted over how important any of that is because we only ever expect the sqlite-backed impl to only be used by a single writer

I'm a touch uncomfortable with anything implementing `Consensus` not being multi-writer safe. Curious what you find here!

File: src/persist/src/sqlite.rs (hunk 4 above)
Comment (danhhz) [reply src/persist/src/sqlite.rs:43]: oh another thought that perhaps makes this last comment irrelevant. my working assumption is that in a single-binary deployment (for those following along, the only place where this is used), then this would share a sqlite instance with the catalog. in which case, we may want `open` to accept a connection instead of a path (the catalog requires that the sqlite instance is opened exclusively)

Comment (ruchirK) [reply src/persist/src/sqlite.rs:43]: hmmm my working assumption was that this would be a separate sqlite instance from the catalog, the way that stash instances are separate sqlite instances from the catalog. I wanted to originally build this on top of stash but it seemed stash really wanted to present as a differential collection (which could be made to work i guess but seemed complex).

File: src/persist/src/sqlite.rs (hunk 2 above)
Comment (ruchirK) [reply src/persist/src/sqlite.rs:117]: I don't think thats correct, because the value of `expected` could be arbitrarily incorrect right? There's no way to guarantee that `expected` is actually <= the current sequence number iiuc

File: src/persist/src/sqlite.rs (hunk 4 above)
Comment (benesch) [reply src/persist/src/sqlite.rs:43]: Discussed with Ruchir in the office, but sharing a SQLite connection among threads leads to `SQLITE_BUSY` exceptions and such. It's generally proven much nicer to let each use manage its own SQLite database(s). `Consensus` feels fundamentally different enough from `Stash` that seems ok and maybe even good to have them separate.

File: src/persist/src/sqlite.rs (hunk 2 above)
Comment (ruchirK) [reply src/persist/src/sqlite.rs:117]: let me rephrase this another way.

Lets say the current seqno is 7.

if someone tries to do a cas with expected = 5, new = 4

we can give them an Err("expected > new") without looking in the database, but the only way to give them a useful value for the current seqno is to look in the database. I felt like it was important to try to return a useful reading of the current seqno as much as possible, but open to changing my mind

File: src/persist/src/sqlite.rs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+#[cfg(test)]
+mod tests {
+    use crate::storage::tests::consensus_impl_test;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn sqlite_consensus() -> Result<(), Error> {
+        let temp_dir = tempfile::tempdir()?;
+        consensus_impl_test(|| {
```
Comment (aljoscha) [inline src/persist/src/sqlite.rs:148]: Nice how you worked this out before! :raised_hands:

**Review by aljoscha — COMMENTED** [review]
I looked at all of it, but don't have useful comments beyond what Dan already mentioned.

File: src/persist/src/sqlite.rs (hunk 5 above)
Comment (danhhz) [reply src/persist/src/sqlite.rs:26]: dunno about sqlite, you'd have to check, but in e.g. postgres if you don't specify a primary key, it invents a meaningless one. then your unique index ends up as a second map that essentially points at the meaningless (hidden) primary key. if you explictly declare a meaningful primary key, you can(i think) save the second map here

as for what the primary key is, for the current state where we're not yet keeping history, shard is a fine primary key. but when we add scan and truncate, i think we'll want the both the uniqueness constraint and pk to be (shard, seqno). if you insist, i suppose i'm okay with starting with the first one and changing to the second later, but given that a large part of this work is de-risking the eventual api, I think it's more useful to start with the second

File: src/persist/src/sqlite.rs (hunk 4 above)
Comment (danhhz) [reply src/persist/src/sqlite.rs:43]: i'm convinced! separate sgtm

**Comment by ruchirK** [comment]
Ok this is ready for a look! this is my best understanding of the approach we'd want to take on a sqlite/pg compatible system. Some notes:

1. This is exactly one request on the uncontended fast path - a insert statement with two select subqueries that should use the index built for the primary key.
2. If the insert fails, this issues a second request to get the current sequence number and data. I didn't see any way around that (the `RETURNING` clause only returns data for successfully inserted/updated rows). I'll keep looking but my
expectation is that the contended case will be relatively rare and its ok for that to be slow.
3. This PR adds a call to truncate which is used within `compare_and_set` to bound the space usage. I felt that was a friendly thing to do for the time being as otherwise the space usage grows without bound and I wasn't sure how aggressively this implementation would be used before we added scan and truncate into the official Consensus API. I don't have any strong preferences to keep it this way and am happy to remove it if we feel comfortable with the space growing while we prototype.

File: src/persist/src/sqlite.rs (hunk 4 above)
Comment (ruchirK) [reply src/persist/src/sqlite.rs]: did the AsRef<Path> thing!

File: src/persist/src/sqlite.rs (hunk 1 above)
Comment (danhhz) [reply src/persist/src/sqlite.rs:83]: ping on this?

File: src/persist/src/sqlite.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                "$shard": self.shard,
+                "$sequence_number": new.seqno.0,
+                "$data": new.data,
+            })?
+        };
+
+        if result == 1 {
+            Ok(Ok(()))
+        } else {
+            let current = self.head(deadline).await?;
```
Comment (danhhz) [inline src/persist/src/sqlite.rs:171]: add a (very quick) note please about why the race condition of having this in a follow txn is safe

**Review by danhhz — APPROVED** [review]
great stuff! LGTM modulo these two last things

File: src/persist/src/sqlite.rs (hunk 7 above)
Comment (ruchirK) [reply src/persist/src/sqlite.rs:177]: done!

File: src/persist/src/sqlite.rs (hunk 1 above)
Comment (ruchirK) [reply src/persist/src/sqlite.rs:83]: I don't think so unfortunately. I searched both the `rusqlite` and `sqlite` docs for `timeout` and `deadline` and the best I could find was https://docs.rs/rusqlite/latest/rusqlite/struct.Connection.html#method.busy_timeout which isn't really clear on what exactly it does. Like, on encountering busy, does it just wait N ms, and then retry and then if its still busy does it give up and exit, or does wake up and retry multiple times during the N ms?

Either way, it didn't seem obvious how to support our requested deadline semantics here, so opted to punt on this in this PR

**Comment by ruchirK** [comment]
TFTR! Merging on green!
